### PR TITLE
Add halium9 support to hfd

### PR DESCRIPTION
**Pull request by erfanoabdi**
This is complete version of https://github.com/ubports/hfd-service/pull/8

### CODE REVIEW

File: src/vibrator-hidl.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#include <unistd.h>
+
+namespace hfd {
+
+bool VibratorHIDL::usable() {
+    return u_hardware_vibrator_new() != NULL;
+}
+
+VibratorHIDL::VibratorHIDL(): Vibrator() {
+    m_device = u_hardware_vibrator_new();
```
Comment (NotKit) [inline src/vibrator-hidl.cpp:32]: Will this crash if Android-side library is missing?

Comment (erfanoabdi) [reply src/vibrator-hidl.cpp:32]: didn't test that

**Comment by Flohack74** [comment]
Can we get this resolved, it seems to block a few Halium 9 ports...

**Comment by kcxt** [comment]
> Can we get this resolved, it seems to block a few Halium 9 ports...

The OnePlus 6 is dependant on this for haptics, would be good to get this fixed.

**Comment by SevralT** [comment]
> Can we get this resolved, it seems to block a few Halium 9 ports...

Would be good to get this merged, because OnePlus 6(T) dependant on this, to fix haptics.

**Comment by Flohack74** [comment]
Yes we know it will unblock a lot of devices, but please have patience, its not fully tested, and currently we cannot build rootfs anyways because of ongoing merges of the Qt 5.12 upgrades. It will be finished by end of week, as sson as we can build again, we can start the remaining tests.

**Comment by kcxt** [comment]
Given concerns about this service working on devices without halium following this PR, would it not be sensible to split hfd-service? Either into `-mainline` and `-halium` packages, and then setting build flags to pick which package gets built (or even having 2 separate repos)?

File: src/hardware/hfd_api.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * Authored by: Erfan Abdi <erfangplus@gmail.com>
+ */
+
+// C APIs
+#include <hfd/hardware/light.h>
+#include <hfd/hardware/vibrator.h>
+#include <hardware/android_hw_module.h>
+
+// Light
+IMPLEMENT_FUNCTION0(
```
Comment (NotKit) [inline src/hardware/hfd_api.cpp:25]: I yet have to test that, but I think it should be IMPLEMENT_CTOR0 instead, which allows to return nullptr if implementation part that is linked against bionic (inside Android image) is missing.

File: src/hardware/hfd_api.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+u_hardware_light_new);
+
+IMPLEMENT_VOID_FUNCTION3(
+u_hardware_light_setLight,
+UHardwareLight,
+light_state_t,
+char const*);
+
+// Vibrator
+IMPLEMENT_FUNCTION0(
```
Comment (NotKit) [inline src/hardware/hfd_api.cpp:36]: Same as for u_hardware_light_new

File: src/vibrator-hidl.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "vibrator-hidl.h"
+
+#include <iostream>
+#include <unistd.h>
+
+namespace hfd {
+
+bool VibratorHIDL::usable() {
+    return u_hardware_vibrator_new() != NULL;
```
Comment (NotKit) [inline src/vibrator-hidl.cpp:28]: We should avoid calling u_hardware_vibrator_new() twice and leaking its result if it succeeds during usable() check. I'm not sure of best solution, but how about having try_create() static method that would just return a smart pointer to VibratorHIDL if succeds?

**Comment by Flohack74** [comment]
Ok can we pull this in for OTA-17? A lot of stuff is blocked by that...

**Comment by UniversalSuperBox** [comment]
This now conflicts with `hfd-service` in xenial. Is it likely that this will be ready for merge in a week or two, or should I remove it from OTA-17?

**Comment by CactiChameleon9** [comment]
Sorry to nudge, but is this being merged?

**Comment by SevralT** [comment]
@erfanoabdi Hello! Will you fix merge conflict or no??

**Comment by erfanoabdi** [comment]
i rather this doesn't get merged at all, me or someone else should rewrite it with gbinder

**Pull request closed without merging**